Point out the type of associated types in every method call of iterator chains

estebank opened the pull request:
Partially address #105184 by pointing out the type of associated types in every method call of iterator chains:

```
note: the expression is of type `Map<std::slice::Iter<'_, {integer}>, [closure@src/test/ui/iterators/invalid-iterator-chain.rs:12:18: 12:21]>`
    --> src/test/ui/iterators/invalid-iterator-chain.rs:12:14
     |
10   |         vec![0, 1]
     |         ---------- this expression has type `Vec<{integer}>`
11   |             .iter()
     |              ------ associated type `std::iter::Iterator::Item` is `&{integer}` here
12   |             .map(|x| { x; })
     |              ^^^^^^^^^^^^^^^ associated type `std::iter::Iterator::Item` is `()` here
```

We also reduce the number of impls we mention when any of the candidates is an "exact match". This benefits the output of cases with numerics greatly.

Outstanding work would be to provide a structured suggestion for appropriate changes, like in this case detecting the spurious `;` in the closure.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3623926734/jobs/6110339186) [(plain)](https://github.com/rust-lang/rust/commit/179525c4377de8509f9f0f00773f848447bbe0f0/checks/9899902046/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
* highest error code: E0790
Found 506 error codes
Found 0 error(s) in error codes
Done!
tidy error: /checkout/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs: too many lines (3088) (add `// ignore-tidy-filelength` to the file to suppress this error)
some tidy checks failed
Build completed unsuccessfully in 0:00:12
```

</details>

compiler-errors commented on compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:
```diff
@@ -2837,26 +2846,172 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
+                                        }
+                                        self.var_for_def(span, param)
+                                    });
+                                let ty_var = self.infcx.next_ty_var(origin);
+                                let trait_ref = ty::Binder::dummy(ty::PredicateKind::Clause(
+                                    ty::Clause::Projection(ty::ProjectionPredicate {
+                                        projection_ty: ty::ProjectionTy { substs, ..*proj },
+                                        term: ty_var.into(),
+                                    }),
+                                ));
+                                ocx.register_obligation(Obligation::misc(
+                                    self.tcx,
+                                    span,
+                                    expr.hir_id,
+                                    param_env,
+                                    trait_ref,
+                                ));
```
You could use `fully_normalize` here. Or just `ocx.normalize` on the projection.

estebank replied:
I'm not so sure I can use `fully_normalize` because the obligation *must* be included, but I could use `ocx.normalize` though.

compiler-errors replied:
I'm not sure if I understand -- you could forgo `ocx` and avoid creating an inference variable if you pass the projection (turned into a ty) into `fully_normalize`. Unless I'm misunderstanding what you're doing here, you're just trying to normalize the projection into a concrete value, right?

estebank replied:
my bad, your snippet made sense

compiler-errors commented on compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:
```diff
@@ -2837,26 +2846,172 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
+                                            ty::GenericParamDefKind::Lifetime
+                                            | ty::GenericParamDefKind::Const { .. } => {}
+                                        }
+                                        self.var_for_def(span, param)
```
Don't we want to use the proj's old substition here? Using a type/const var and normalizing it under a totally different param-env is probably going to cause some strange issues.

estebank replied:
So I originally had `i32: Sum<()>`, from that I got `S: Sum<A>`. From that I'm getting `<Self as Iterator>::Item` where `Self == S`, which I want to see, given `S: Sum<A>`, what does `A` have to be for `i32: Sum<_>`. That's why I'm getting a new inference type.

compiler-errors replied:
Oh shoot, I thought `var_for_def` returned a `ty::Param` for some reason, lol.

compiler-errors commented on compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:
```diff
@@ -2837,26 +2846,172 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
+                                let ty_var = self.infcx.next_ty_var(origin);
+                                let trait_ref = ty::Binder::dummy(ty::PredicateKind::Clause(
+                                    ty::Clause::Projection(ty::ProjectionPredicate {
+                                        projection_ty: ty::ProjectionTy { substs, ..*proj },
```
nit: better to be explicit here

```suggestion
                                        projection_ty: ty::ProjectionTy { item_def_id: proj.item_def_id, substs },
```

compiler-errors commented on compiler/rustc_middle/src/ty/_match.rs:
```diff
@@ -131,3 +131,72 @@ impl<'tcx> TypeRelation<'tcx> for Match<'tcx> {
+        Ok(a)
+    }
+    fn tys(&mut self, a: Ty<'tcx>, b: Ty<'tcx>) -> RelateResult<'tcx, Ty<'tcx>> {
+        if a == b || matches!(a.kind(), ty::Infer(_)) || matches!(b.kind(), ty::Infer(_)) {
```
This should probably use `InferCtxt::can_eq` instead of these hard-coded cases and `Ty` equality. We'll need to pass in an `InferCtxt`, but it should be more robust that way.

estebank replied:
rustc_middle doesn't depend on rustc_infer :-/

compiler-errors replied:
(or even in `compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs` itself, so people aren't tempted to use it without understanding what this relation does, hehe)

estebank replied:
Done!

estebank replied:
Ah, right. The matches here I added to avoid the following super_relate_tys from blowing up and not to make them "equal". :-/

compiler-errors replied:
haha, that's because we fail to relate a concrete type against an *integer* inference variable. You probably should be using `super_combine_tys` instead of `super_relate_tys` to handle this case, and putting it inside of a probe. For example: `Err(Sorts(ExpectedFound { expected: A, found: {integer} }))`

But it seems like there's a lot of back and forth with my suggestions causing ICEs, so if you don't want to fix this then I'll do it as a follow-up, since `{integer} == ConcreteType` is possibly going to lead to some weird diagnostics.

compiler-errors commented on compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:
```diff
@@ -2837,26 +2846,187 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
+                                        }
+                                        self.var_for_def(span, param)
+                                    });
+                                // This will hold the resolved type of the associated type, if the
+                                // current expression implements the trait that associated type is
+                                // in. For example, this would be what `Iterator::Item` is here.
+                                let ty_var = self.infcx.next_ty_var(origin);
+                                // This corresponds to `<ExprTy as Iterator>::Item = _`.
+                                let trait_ref = ty::Binder::dummy(ty::PredicateKind::Clause(
+                                    ty::Clause::Projection(ty::ProjectionPredicate {
+                                        projection_ty: ty::ProjectionTy {
+                                            substs,
+                                            item_def_id: proj.item_def_id,
+                                        },
+                                        term: ty_var.into(),
+                                    }),
+                                ));
+                                // Add `<ExprTy as Iterator>::Item = _` obligation.
+                                ocx.register_obligation(Obligation::misc(
+                                    self.tcx,
+                                    span,
+                                    expr.hir_id,
+                                    param_env,
+                                    trait_ref,
+                                ));
+                                if ocx.select_where_possible().is_empty() {
```
Something like:

```suggestion
                                let normalized_ty = fully_normalize(self, /* other fields */, tcx.mk_projection(ty::ProjectionTy { .. }));
                                if let Ok(..) = normalized_ty {
```

compiler-errors commented on compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:
```diff
@@ -289,16 +293,16 @@ pub trait TypeErrCtxtExt<'tcx> {
         &self,
         err: &mut Diagnostic,
-        predicate: &T,
+        predicate: Option<ty::Predicate<'tcx>>,
+        predicate_fallback: String,
```
#105342 should fix this strange wart by making the `predicate` argument always impl `ToPredicate`.

estebank replied:
I wanted to, but I would have to add a variant for the case of a Ty Predicate :-/

compiler-errors replied:
The PR I linked avoids that by not noting the cause code for `Obligation<Ty<'tcx>>` -- the only case that happens is during opaque type overflow, which only happens during codegen, and is almost certainly not associated with a useful obligation cause.

estebank replied:
Ah, great. If that works great (but I was seeing real cases of type overflows with a need to display useful obligation causes. I'll rebase on top of yours after it lands.

compiler-errors replied:
`<T as ToPredicate>::to_predicate` should give you back a ty::Predicate, not a `&T`?

estebank commented on compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:
```diff
@@ -2837,26 +2846,187 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
+                        // instead of `Vec<_>`.
+                        if let Some(ty) = typeck_results
+                            .expr_ty_opt(expr)
+                            .or_else(|| typeck_results.expr_ty_adjusted_opt(expr))
```
Probably not :)

bors commented:
:umbrella: The latest upstream changes (presumably #105365) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

compiler-errors committed:
Provide associated type information in method chains

When encountering an unmet obligation that affects a method chain, like
in iterator chains where one of the links has the wrong associated
type, we point at every method call and mention their evaluated
associated type at that point to give context to the user of where
expectations diverged from the code as written.

```
note: the expression is of type `Map<std::slice::Iter<'_, {integer}>, [closure@$DIR/invalid-iterator-chain.rs:12:18: 12:21]>`
  --> $DIR/invalid-iterator-chain.rs:12:14
   |
LL |         vec![0, 1]
   |         ---------- this expression has type `Vec<{integer}>`
LL |             .iter()
   |              ------ associated type `std::iter::Iterator::Item` is `&{integer}` here
LL |             .map(|x| { x; })
   |              ^^^^^^^^^^^^^^^ associated type `std::iter::Iterator::Item` is `()` here
```

estebank committed:
Expand iterator chain test

estebank committed:
Mention only assoc types changes

estebank committed:
Account for method call chains split across multiple bindings

estebank committed:
Remove mention of "assoc type" in label as it is already in the `note` message

estebank committed:
Add label to method chains where assoc type remains the same

estebank committed:
Only point at methods that might be relevant

estebank committed:
fix rebase

estebank committed:
Use `with_forced_trimmed_paths`

oli-obk commented on src/test/ui/iterators/invalid-iterator-chain.stderr:
```diff
@@ -0,0 +1,199 @@
+LL | |             .map(|x| x as f64)
+LL | |             .filter(|x| *x > 0.0)
+LL | |             .map(|x| { x + 1.0 })
+   | |_________________________________^ value of type `i32` cannot be made by summing a `std::iter::Iterator<Item=f64>`
```
Should use trimmed paths here, too?

estebank replied:
It's tricky because this is coming now from on_unimplemented, and we explicitly want the long paths there for filtering and we don't currently have a way of offering the long path for filtering while also replacing the formatting string with the short paths. :-/

I think it is fine for now.

oli-obk commented:
Could be a good opportunity to move the entire ObligationCauseCode::FunctionArgumentObligation arm into a new module

oli-obk commented on compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:
```diff
@@ -2838,26 +2841,271 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
+                    {
+                        let mut c = CollectAllMismatches {
+                            infcx: self.infcx,
+                            param_env: param_env,
```
```suggestion
                            param_env,
```

oli-obk commented on compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:
```diff
@@ -2838,26 +2841,271 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
+                                predicate
+                            )) = predicate.kind().skip_binder()
+                        {
+                            if let Ok(_) = c.relate(trait_pred, predicate) {
```
Should probably keep the binders aroundfor both and compare under binders

oli-obk commented on compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:
```diff
@@ -2838,26 +2841,271 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
+                                    }),
+                                ));
+                                // Add `<ExprTy as Iterator>::Item = _` obligation.
+                                ocx.register_obligation(Obligation::misc(
```
I think you can just use a dummy obligation cause

estebank replied:
Isn't that what `Obligation::misc` uses under the covers?

oli-obk commented on compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:
```diff
@@ -2838,26 +2841,271 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
+                            }
+                        }
+                    }
+                    let point_at_chain = |expr: &hir::Expr<'_>| {
```
This should be a separate function. Way too long to be called later.

estebank replied:
Moved to its own method, I'll file a follow up PR to split the trait through multiple files to make it more manageable.

estebank committed:
Point at method call when it is the source of the bound error

estebank committed:
Move logic to their own methods

oli-obk commented on compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:
```diff
@@ -3099,6 +3090,295 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
+                    param_env,
+                    errors: vec![],
+                };
+                if let ty::PredicateKind::Clause(ty::Clause::Trait(
+                        predicate
+                    )) = predicate.kind().skip_binder()
+                {
+                    if let Ok(_) = c.relate(trait_pred, predicate) {
+                        type_diffs = c.errors;
+                    }
+                }
```
This is still comparing with skipping binders, you can compare the bound types by mapping the binders instead of skipping them

estebank replied:
done

estebank committed:
Do not `skip_binder`s

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit f4ed2d1cca0106eb4aec99b61670ecac378ca087 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk f4ed2d1cca0106eb4aec99b61670ecac378ca087 -->
<!-- homu: {"type":"Approved","sha":"f4ed2d1cca0106eb4aec99b61670ecac378ca087","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
